fb3p/pytorch-gaudi
# PyTorch Habana Gaudi enablement patch

The following changes are part of this patch -
  - Add Habana as a backend device type
  - Add HabanaTensorId as a DispatchKey for Habana backend
  - Adding support to recognize prim::HabanaFusedOp as an operator
  - Add allgather_base to ProcessGroupHCL
  - Update distributed test
  - Adding support to call autograd threshold for backward pass
    in lieu of symbolic implementation
  - After making the changes, verified by running test_jit.py
    test suite from tests dir
  - Add support for reduce collective
  - Temporary change to accomodate the lack of Integer allreduce
    support in HCL
  - Ensure convolution op in TorchScript is a differentiable op
    1. Added support in graph executor to ensure that TensorTypes
       are of the same device as IValue tensors
    2. Added support for checking if Tensor, types and other methods
       are on habana by defining helper functions
    3. Added checks in autodiff to ensure convolution is differentiable
       only for HPU
    4. Added symbolic script implementation for convolution and _convolution
  - Ensure BatchNorm is part of differentiable graph in TorchScript
    1. Enable BatchNorm implementation in symbolic script
    2. Added check to include batchnorm to be part of differentiable
       graph for HPU
